### PR TITLE
Add missing r_vector_fini implementation ##util

### PR DESCRIPTION
**Pull request by radare**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

RVector API lacks .fini() method, also the tests are badly written and that fini needed to be tested and fixed

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
seems broken though

File: libr/include/r_vector.h [hunk 1]
```diff
@@ -74,6 +74,11 @@ static inline bool r_vector_empty(const RVector *vec) {
 	return vec->len == 0;
 }
 
+// backward compat
+R_DEPRECATE static inline void r_vector_clear(RVector *vec) {
```
Comment (ret2libc) [inline libr/include/r_vector.h:78]: I don't agree on this "deprecate" thing. "clear" is a very common method name for vectors, more than "fini" (which to me is actually confusing). So I would keep this, without the R_DEPRECATE. @thestr4ng3r ?

**Comment by radare** [comment]
Broken and aproved?

File: libr/util/vector.c [hunk 2]
```diff
@@ -56,15 +54,16 @@ static void vector_free_elems(RVector *vec) {
...
 }
 
-R_API void r_vector_clear(RVector *vec) {
-	vector_free_elems (vec);
-	R_FREE (vec->a);
-	vec->capacity = 0;
+R_API void r_vector_fini(RVector *vec) {
+	if (vec) {
+		vector_free_elems (vec);
+		free (vec->a);
```
Comment (ret2libc) [inline libr/util/vector.c:60]: R_FREE maybe?

**Comment by radare** [comment]
i see java and c++ have this clear method for the vector class, and i respect that, the thing is that init/fini is more about the liifetime of the instance. funnily in this implementation deinitializing the instance is  the same as clearing its conteents. But thats an implementatiion specific feature. i can add a method instead of this static inline alias between them to make that clear if you want

File: libr/util/vector.c (hunk 2 above)
Comment (thestr4ng3r) [reply libr/util/vector.c:60]: Yes actually R_FREE is very important.

File: libr/include/r_vector.h (hunk 1 above)
Comment (thestr4ng3r) [reply libr/include/r_vector.h:78]: agreed

**Comment by radare** [comment]
Wait i have tests to push

**Comment by radare** [comment]
check the test and imho its ready to go. pls review and aprove if lgty

File: test/unit/test_vector.c [hunk 3]
```diff
@@ -61,6 +61,33 @@ static bool _init_test_pvector2(RPVector *v, size_t len, size_t padding) {
 	mu_assert ("init_test_pvector2", _r); \
 }
 
+static bool test_vector_fini() {
+	RVector v;
+	r_vector_init (&v, sizeof (void *), NULL, NULL);
+	r_vector_push (&v, &v);
+	mu_assert_eq_fmt (v.elem_size, 8UL, "init elem_size", "%lu");
+	mu_assert_eq_fmt (v.len, 1UL, "init len", "%lu");
+	mu_assert_eq_fmt ((void*)(size_t)!v.a, NULL, "init a", "%p");
```
Comment (ret2libc) [inline test/unit/test_vector.c:70]: `mu_assert_notnull (v.a, "init a");`

File: test/unit/test_vector.c [hunk 4]
```diff
@@ -61,6 +61,33 @@ static bool _init_test_pvector2(RPVector *v, size_t len, size_t padding) {
 	mu_assert ("init_test_pvector2", _r); \
 }
 
+static bool test_vector_fini() {
+	RVector v;
+	r_vector_init (&v, sizeof (void *), NULL, NULL);
+	r_vector_push (&v, &v);
+	mu_assert_eq_fmt (v.elem_size, 8UL, "init elem_size", "%lu");
+	mu_assert_eq_fmt (v.len, 1UL, "init len", "%lu");
```
Comment (ret2libc) [inline test/unit/test_vector.c:69]: `mu_assert_eq (v.len, 1, "init len");`

Comment (ret2libc) [reply test/unit/test_vector.c:69]: same for all the others "%lu" formats

File: test/unit/test_vector.c [hunk 5]
```diff
@@ -61,6 +61,33 @@ static bool _init_test_pvector2(RPVector *v, size_t len, size_t padding) {
...
+	mu_assert_eq_fmt ((void*)(size_t)!v.a, NULL, "init a", "%p");
+	mu_assert_eq_fmt (v.capacity, 4UL, "init capacity", "%lu");
+	mu_assert_eq_fmt (v.free, (void *)0, "init free", "%p");
```
Comment (ret2libc) [inline test/unit/test_vector.c:72]: `mu_assert_null (v.free, "init free");`

File: test/unit/test_vector.c (hunk 3 above)
Comment (ret2libc) [reply test/unit/test_vector.c:70]: use `mu_assert_null`/`mu_assert_notnull` for all the other pointers operation.

File: test/unit/test_vector.c (hunk 4 above)
Comment (radare) [reply test/unit/test_vector.c:69]: done

File: test/unit/test_vector.c (hunk 3 above)
Comment (radare) [reply test/unit/test_vector.c:70]: all those tests are not mine. but ill fix them

Comment (ret2libc) [reply test/unit/test_vector.c:70]: yeah i wasn't looking at the others actually. just saw yours. they could have been better as well :/

**Comment by radare** [comment]
ok ive pushed more fixed tests but there are more still

**Comment by radare** [comment]
review again, fini was not properly fixed

File: test/unit/test_vector.c [hunk 6]
```diff
@@ -723,51 +750,51 @@ static bool test_pvector_insert() {
...
-	mu_assert_eq_fmt (v.v.len, 5UL, "insert (end) => len", "%lu");
-	mu_assert_eq_fmt (e, (void *)1337, "insert (end) => content at returned ptr", "%p");
-	mu_assert_eq_fmt (*((void **)r_vector_index_ptr (&v.v, 0)), (void *)0, "insert (end) => old content", "%p");
-	mu_assert_eq_fmt (*((void **)r_vector_index_ptr (&v.v, 1)), (void *)1, "insert (end) => old content", "%p");
-	mu_assert_eq_fmt (*((void **)r_vector_index_ptr (&v.v, 2)), (void *)2, "insert (end) => old content", "%p");
-	mu_assert_eq_fmt (*((void **)r_vector_index_ptr (&v.v, 3)), (void *)3, "insert (end) => old content", "%p");
-	mu_assert_eq_fmt (*((void **)r_vector_index_ptr (&v.v, 4)), (void *)1337, "insert (end) => content", "%p");
+	mu_assert_eq (v.v.len, 5UL, "insert (end) => len");
+	mu_assert_eq (e, (void *)1337, "insert (end) => content at returned ptr");
+	mu_assert_eq (*((void **)r_vector_index_ptr (&v.v, 0)), (void *)0, "insert (end) => old content");
```
Comment (ret2libc) [inline test/unit/test_vector.c:780]: just FYI (you can keep it this way here) there is also `mu_assert_ptreq` and `mu_assert_ptrneq`.

File: libr/util/vector.c [hunk 7]
```diff
@@ -56,15 +54,22 @@ static void vector_free_elems(RVector *vec) {
...
+		vector_free_elems (vec);
+		R_FREE (vec->a);
+		vec->capacity = 0;
+	}
 }
 
 R_API void r_vector_free(RVector *vec) {
-	vector_free_elems (vec);
-	free (vec->a);
+	r_vector_fini (vec);
```
Comment (thestr4ng3r) [inline libr/util/vector.c:72]: Will null-deref. Imo only check for null once in `_free()` and not in `_clear()` or `_fini()`.

File: test/unit/test_vector.c (hunk 3 above)
Comment (radare) [reply test/unit/test_vector.c]: i just followed the style, but now its all fixed

File: libr/util/vector.c (hunk 7 above)
Comment (radare) [reply libr/util/vector.c:75]: fixed. pls review again

**Pull request merged**